[READY] Adds Pulsating Tumor Ruin and Lavaland Elites

IndieanaJones opened the pull request:
## About The Pull Request

This PR adds the a pulsating tumor ruin to Lavaland, and along with it a slew of new combat-based content for both novice and experienced miners.  With the very minimal explanation out of the way, let's get into the thick of it:

**THE PULSATING TUMOR**
The pulsating tumor is an odd land deformity which spawns as a ruin on Lavaland.  One is guaranteed to spawn in at round start, but it's possible more than one could spawn in.  The tumor itself also has a GPS component, which is listed as 'Menacing Signal'.  At first glance, the tumor doesn't appear to be much, but it's when you get into the interactions with it that the real girth of content comes to light...

**BASE INTERACTION**
Upon interacting with the tumor, an ominous message will be put into in chat.  After about 3 seconds, the tumor will select one specially-designed mob out of a list to spawn as the tumor's "child".  Said mobs are dubbed "Lavaland Elites", and are much harder to fight than common spawn on lavaland, but less lethal than megafauna.
Upon the lavaland elite spawning in, a barrier will be formed around the tumor, similar to the Hierophant's.  This barrier only affects the player who interacted with the tumor (henceforth dubbed the activator) and the lavaland elite, letting anything go through it but specifically those two.  Should either manage to teleport outside of the arena, they will be teleported back on top of the tumor after at most 10 seconds.  From here, the player and the lavaland elite must fight to the death, where whoever loses determining what happens next.

**ACTIVATOR WINS**
Should the activator win the fight, the barrier surrounding the tumor will vanish and the tumor will make a noise identical to the tendrils and drop a necropolis chest.  The tumor then vanishes, and is completely gone from the game.  The corpse of the elite remains and depending from mob to mob, can be butchered to achieve whatever results.  From this point, all interaction is basically over.

**ELITE WINS**
Should the lavaland elite win, it is healed to full, and the barrier then goes down.  The lavaland elite can roam freely where ever it pleases, and can also be killed during this time.  Should this occur, the tumor will vanish as it would have should the activator won, but no loot will be dropped.  The tumor also becomes interactable once more during this time, and after a brief warning and time delay, will re-summon the elite back to the tumor, heal it fully, and put up the barrier, basically going right back to as if it were just activated.  The same rules apply to this fight as the first, with it being a deathmatch between activator and elite and the same results are achieved when either side wins.

So far this is pretty basic, a ruin which lets you summon middle-ground mob between normal and mega fauna for some basic loot.  But, this is not the only interaction the glowing tumor has available to it.

**BOOSTING THE PULSATING TUMOR**
Should a stabilized legion core be used on the tumor before first activation, it will be dropped into the tumor, boosting it's power.  This makes it stronger, empowering the mob which comes out... with sentience.

**BOOSTED ELITE**
Upon interacting with a boosted tumor, all ghosts with the sentience role will be polled on if they'd like to be a lavaland elite.  If no ghost agrees, the tumor will not activate, and the player will be allowed to try again some other time.  Should ghosts say yes, they'll be added to a pool of possible candidates and one will be chosen randomly.  Upon being chosen, the ghost will receive a sound notification and prompt on what is about to occur, which lasts about 10 seconds.  Once this happens, the tumor will put out a unique message, warning the activator that an elite is about to spawn.  After the 10 seconds of grace, the mob is spawned in with the ghost in control of it, and the fight advances normally.

**DIFFERENCES FROM REGULAR TUMOR**
The lavaland elite, outside of being player-controlled, is 100% the same.  Its only the outcomes from this point which change, which are mentioned below:

**ACTIVATOR WINS BOOSTED**
Should the activator win against a boosted elite, the tumor will give a necropolis chest like before, but also has a 50%-50% chance of spawning a unique loot drop based on the elite fought or a tumor shard.  The tumor shard functions very similarly to a lazarus injector, only that it only works on sentient lavaland elites.  Using it on a corpse with a ghost in it revives the elite, makes it loyal to the user and cuts its health in half.  It also adds an extra part to their description to differentiate them from regular elites, and lets them be revived via lazarus injector.   After this point, all interactions are over.

**ELITE WINS BOOSTED**
Should the elite win boosted, the same as a regular elite win occurs, except the elite is notified that it is free but can be called back if someone interacts with the tumor.  Along with this, it's health is cut in half, but it can heal by standing on the tumor.  Note, this is the only way it can heal, kills and such do it give any health.

**SOME THINGS ABOUT ELITES**
- Cannot be made sentient or lazurus'd, due to having SENTIENCE_BOSS
- Only deal 5 structure damage, and cannot break walls or rwalls.  They can, however, break mineral walls in one hit.  This is to prevent sentient elites from taking apart the mining station, which would be very un-fun.
- Will attack targets until they are dead
- Are only a part of the boss faction, meaning they are on the same team as megafauna, not not regular fauna.
- Cannot attack those on their own faction (which includes themselves)
- Can select their attacks with buttons on the top left of the screen
- Cannot talk (unless specifically enabled to, will likely use for the colossus-based elite)
- Can see in the dark

**CURRENT ELITES**
Goliath Broodmother:
- 800 HP
- 30 Melee
- 2 Speed
- When below 50% health, random tentacles will begin spawning near the broodmother.  At 25% health, this effect is doubled.  Melee attacks have 30 armor piercing.
 ATTACKS -
  Tentacle Patch: Summon a large patch of tentacles by the target (combination of 3 x 3 and 5 x 5 plus-shape.)  These tentacles have less stun time than a normal goliath's, but deal more damage.
  Children Spawn: Summon 2 of its children to its location.  These are baby goliaths, with the following stats:
    - 20 Health
    - 5 Damage
    - 2 Speed
    - When killed, explodes in a fiery explosion
    - At range, will apply a tentacle to the square under their target.  Has a cooldown of 4 seconds.
  The broodmother at max can have 8 of them out at a time.
  Rage: Increases in speed for 6.5 seconds, with a cooldown of 7 seconds.
  Call Children: All active children will be summoned around the broodmother upon use.

Unique Loot: Broodmother Tongue
Functions as a crusher trophy.  Gives a 10% chance to summon a patch of tentacles at the target on mark detonation.
  
Pandora:
- 800 HP
- 15 Melee
- 4 Speed
- All attacks have a cooldown of 1 second sans Blast Line, and is decreased by 25% at 50% HP and 25% HP.  Blast line has half the normal cooldown of other attacks.
ATTACKS -
  Blast Line: Fires a line of Hierophant blasts in either a cardinal or diagonal direction.
  Magic Box: Surrounds the target with Hierophant blasts, leaving a 3x3 square in the center.  Punishes mindless running away from Pandora, along with fitting the box theme.
  Teleport: Almost identical to Hierophant's teleport, deals damage in a 3x3 at the initial location and at the end location.
   AOE Blast: Does a 5x5 blast, stretching out from the center.  Unlike Hierophant, this attack is one done at a range.

Unique Loot: Hope
Functions as a medal.  Gives a positive moodlet when pinned on your uniform.

Legionnaire:
 - 800 HP
 - 30 Melee
 - 1 Speed
 - Legionnaire's skull initially does 10 damage, but ramps up to 20 at 50%, and 30 at 25%.
  ATTACKS:
  Legionnaire Charge - Charges forward 4 tiles, throwing anyone in its path forward and dealing a whopping 50 damage.  Attack is telegraphed for half a second, and legionnaire can slightly adjust the travel path when charging.  Don't stand too close!
  Skull Detach - The legionnaire's skull detaches, fighting as its own entity.  Will chase targets and deal 10 damage to them, ramping up as the legionnaire's health goes down.
  Bone Bonfire - If no bone bonfire is currently active, places a bonfire on legionnaire's current tile.  When used again, the legionnaire will swap places with the bonfire.  The bonfire can be destroyed and walked over, and using the ability when standing on the active bonfire destroys it.  If a flammable creature walks over the bonfire, they will be set on fire.
  Smoke Spew - Spawns a smoke cloud around Legionnaire.  If its head is currently detached, the cloud will be spawned at the head's location instead.

Unique Loot: Legionnaire Spine
Functions as a trophy for the kinetic crusher.  Gives the crusher a 20% chance to summon a legion head to attack the target for 10 seconds on mark detonation.

Herald:
- 800 HP
- 20 Melee
- 4 Speed
- Once below 50% health, herald's tri-shot and circular shot will fire twice when used.
- Herald's deathbolts don't break structures immediately and only do 15 damage, with less armor penetration (still pretty high)
  ATTACKS:
  Tri-shot - Fires three bolts at the target position.
  Spread-shot - Herald raises it's arms(?) and shakes before firing a deathbolt in every cardinal and diagonal direction at once.  Has longer delay to fire again than tri-bolt.
  Tele-shot - Fires a golden bolt which does no damage.  Once it strikes something, the Herald will be teleported to where it hit.
  Mirror - Summons a mirror which reflects yourself.  The mirror will replicate all your attacks, firing at the same position as you do.  The only effect it doesn't duplicate is doubling it's attack when your health is low.  Tele-bolt will teleport the mirror as it teleports the herald.

Unique Loot: Cloak of the Prophet
A cloak based on the appearance of Herald.  When attacked, has a 10% chance to do a spread of herald deathbolts around the wearer.

Other elites are in development, but I wanted to put this PR before I finished in case major changes to the base code needed to be changed.  As of now, this is 100% merge-friendly, functions as described above, the only thing being the only mobs which can spawn are the mobs above.  The planned elites are:

- Elite watcher
- Elite goldgrub
- Bubbles-based Elite
- Ash Drake-based Elite
- ???

## Why It's Good For The Game

This PR hits several good notes in terms of content the game could use.  For one, its new mining content, two, it's something accessible to both novice and experienced miners, and three, adds a series of interesting new ghost roles.  The content is also miner-choice, as nothing will happen should they not activate the tumor.  It also means it's not a free murderbone ghost role either; the miner needs to procure a stabilized core and willingly put it into the crevice for it to allow a ghost to take over.  They're also less prone to griefing, as they lack the ability to smash things, lack the ability to heal outside of their tumor(which they must remember its location), and cannot operate consoles to leave Lavaland by themselves.  For miners, the mobs will also double as good mining partners, as their attacks also double as mining tools with various uses.  Finally, it functions as robustness practice for miners who don't want to get gibbed for losing a fight, letting them use crevices to train to take on megafauna.

## Changelog
:cl: Indie-ana Jones & Meyhazah
add: New lavaland ruin: Pulsating Tumor
add: New lavaland enemies: Lavaland Elites
add: New Lavaland Elite: Goliath Broodmother
add: New Lavaland Elite: Pandora
add: New Lavaland Elite: Legionnaire
add: New Lavaland Elite: Herald
imageadd: Added lavaland elite sprites and pulsating tumor sprites
imageadd: New lavaland loot sprites
imageadd: Elite ability sprites
/:cl:

IndieanaJones committed:
Adds elite.dm and elite_goliath.dm

IndieanaJones committed:
Add lavaland_surface_elite_crevice.dmm

IndieanaJones committed:
Load elite.dm and elite_goliath.dm

IndieanaJones committed:
Add Glowing Crevice Ruin

IndieanaJones committed:
Add elite_lavaland_monsters.dmi

IndieanaJones committed:
Additional Sprite to artefacts.dmi

IndieanaJones committed:
Updated elite_goliath.dm and elite.dm

IndieanaJones committed:
Add actions_elite.dmi

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite_goliath.dm:
```diff
@@ -0,0 +1,234 @@
+	sleep(20)
+	icon_state = "explosive_shroom_active"
+	armed = 1
+	INVOKE_ASYNC(src, .proc/delete_self)
```
QDEL_IN(src, 1 MINUTE)

IndieanaJones replied:
Resolved in upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite_goliath.dm:
```diff
@@ -0,0 +1,234 @@
+		M.gets_drilled()
+	playsound(src, 'sound/effects/bamf.ogg', 100, 1)
+	icon_state = "explosive_shroom_activate"
+	sleep(20)
```
Don't EVER sleep in Initialize

IndieanaJones replied:
Resolved in upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite_goliath.dm:
```diff
@@ -0,0 +1,234 @@
+/*
```
Convert this to proper autodoc format

IndieanaJones replied:
Sorry, could you explain what you mean here?  I just looked at what was done for the megafauna and copied that.

Qustinnus replied:
Since recently we have been using autodoc. https://github.com/tgstation/tgstation/pull/45468/files heres an example

IndieanaJones replied:
Okay, that seems simple enough.  Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+		
+//Elites can't talk (normally)!
+/mob/living/simple_animal/hostile/asteroid/elite/say(message, bubble_type, var/list/spans = list(), sanitize = TRUE, datum/language/language = null, ignore_spam = FALSE, forced = null)
+	if(can_talk == 1)
```
if(can_talk)

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+	if(can_talk == 1)
+		. = ..()
+		return
+	return 0
```
return FALSE

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+/mob/living/simple_animal/hostile/asteroid/elite/say(message, bubble_type, var/list/spans = list(), sanitize = TRUE, datum/language/language = null, ignore_spam = FALSE, forced = null)
+	if(can_talk == 1)
+		. = ..()
+		return
```
if you're going to use a bool for the second return do it for the first too

IndieanaJones replied:
Resolved in upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				E.faction = list("neutral")
+				E.revive(full_heal = 1, admin_revive = 1)
+				user.visible_message("<span class='notice'>[user] stabs [E] with [src], reviving it.</span>")
+				E.mind.enslave_mind_to_creator(user)
```
E probably doesn't even have a mind

IndieanaJones replied:
Removed in upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+		if(E.stat == DEAD && E.sentience_type == SENTIENCE_BOSS)
+			if(E.key)
+				E.faction = list("neutral")
+				E.revive(full_heal = 1, admin_revive = 1)
```
These are booleans

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				var/obj/effect/temp_visual/heal/H = new /obj/effect/temp_visual/heal(get_turf(src))
+				H.color = "#FF0000"
+	if(myparent)
+		if(myparent.activity == 1 && myparent.activator.stat == DEAD)
```
activity is, I'm assuming a bool. do (myparent.activity)

Qustinnus replied:
nvm its not a bool. Use defines to make flags. This is confusing.

IndieanaJones replied:
Bools added in the upcoming fix, which should hopefully make this much easier to read.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+	name = "glowing crevice"
+	desc = "A glowing, red hole which doesn't seem to have a bottom.  You feel pressured to reach your hand out towards it..."
+	armor = list("melee" = 100, "bullet" = 100, "laser" = 100, "energy" = 100, "bomb" = 100, "bio" = 100, "rad" = 100, "fire" = 100, "acid" = 100)
+	max_integrity = 10000
```
just make it indestructible.

IndieanaJones replied:
Done for upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+	desc = "A glowing, red hole which doesn't seem to have a bottom.  You feel pressured to reach your hand out towards it..."
+	armor = list("melee" = 100, "bullet" = 100, "laser" = 100, "energy" = 100, "bomb" = 100, "bio" = 100, "rad" = 100, "fire" = 100, "acid" = 100)
+	max_integrity = 10000
+	var/activity = 0
```
Make these defines instead of magic numbers.

IndieanaJones replied:
Added in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+	density = FALSE
+	
+/obj/structure/elite_crevice/attack_hand(mob/user)
+	if(activity == 0)
```
use a switch statement with flags

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			var/mob/dead/observer/elitemind = null
+			visible_message("<span class='boldwarning'>The crevice glows.  Your instincts tell you to step back.</span>")
+			activator = user
+			if(boosted == 1)
```
if(boosted)

IndieanaJones replied:
Fixed here and all throughout the code in upcoming update.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+/obj/structure/elite_crevice/attack_hand(mob/user)
+	if(activity == 0)
+		if(istype(user, /mob/living/carbon/human))
+			activity = 1
```
flags

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+					visible_message("<span class='boldwarning'>The stirring sounds increase in volume!</span>")
+					elitemind = pick(candidates)
+					SEND_SOUND(elitemind, sound('sound/effects/magic.ogg'))
+					to_chat(elitemind, "<b>You have been chosen to play as a Lavaland Elite.</b>")
```
make this one to_chat. to_chat spam lags the client

IndieanaJones replied:
Done for upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+					to_chat(elitemind, "<b>Your attacks can be switched using the buttons on the top left of the HUD, and used by clicking on targets or tiles similar to a gun.</b>")
+					to_chat(elitemind, "<b>While the opponent might have an upper hand with  powerful mining equipment and tools, you have great power normally limited by AI mobs.</b>")
+					to_chat(elitemind, "<b>If you want to win, you'll have to use your powers in creative ways to ensure the kill.  It's suggested you try using them all as soon as possible.</b>")
+					to_chat(elitemind, "<b>Should you win, you'll receive extra information regarding what to do after.  Good luck!</b>")
```
Don't call to_chat this many times in a row. Just make a newline with \n

IndieanaJones replied:
Resolved here and in another location in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+					to_chat(elitemind, "<b>While the opponent might have an upper hand with  powerful mining equipment and tools, you have great power normally limited by AI mobs.</b>")
+					to_chat(elitemind, "<b>If you want to win, you'll have to use your powers in creative ways to ensure the kill.  It's suggested you try using them all as soon as possible.</b>")
+					to_chat(elitemind, "<b>Should you win, you'll receive extra information regarding what to do after.  Good luck!</b>")
+					sleep(100)
```
dont do this use timers

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+					sleep(100)
+				else
+					visible_message("<span class='boldwarning'>The stirring stops, and nothing emerges.  Perhaps try again later.</span>")
+					activity = 0
```
flag

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+					activator = null
+					return
+			else
+				sleep(30)
```
timer

IndieanaJones replied:
All sleeps are now timers in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			mychild.myparent = src
+			visible_message("<span class='boldwarning'>[mychild] emerges from the crevice!</span>")
+			playsound(loc,'sound/effects/phasein.ogg', 200, 0, 50, 1, 1)
+			if(boosted == 1)
```
if(boosted)

IndieanaJones replied:
Resolved in upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				visible_message("<span class='boldwarning'>Something within the crevice stirs...</span>")
+				var/list/candidates = pollCandidatesForMob("Do you want to play as a lavaland elite?", ROLE_SENTIENCE, null, ROLE_SENTIENCE, 50, src, POLL_IGNORE_SENTIENCE_POTION)
+				if(candidates.len)
+					visible_message("<span class='boldwarning'>The stirring sounds increase in volume!</span>")
```
this is an audible message and not a visible message.

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			visible_message("<span class='boldwarning'>The crevice glows as your arm enters its radius.  Your instincts tell you to step back.</span>")
+			activator = user
+			INVOKE_ASYNC(src, .proc/arena_trap)
+			if(boosted == 1)
```
if(boosted)

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				mychild.sentience_act()
+			INVOKE_ASYNC(src, .proc/arena_trap)
+			return
+	if(activity == 2)
```
more flag stuff. I'm gonna stop commenting this, just make sure u do it.

IndieanaJones replied:
Leave it to me, captain.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				SEND_SOUND(mychild, sound('sound/effects/magic.ogg'))
+				to_chat(mychild, "<b>Someone has activated your crevice.  You will be returned to fight shortly, get ready!</b>")
+				sleep(40)
+			sleep(30)
```
timer

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			if(boosted == 1)
+				SEND_SOUND(mychild, sound('sound/effects/magic.ogg'))
+				to_chat(mychild, "<b>Someone has activated your crevice.  You will be returned to fight shortly, get ready!</b>")
+				sleep(40)
```
timer

IndieanaJones replied:
Fix for upcoming done.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			visible_message("<span class='boldwarning'>[mychild] emerges from the crevice!</span>")
+			playsound(loc,'sound/effects/phasein.ogg', 200, 0, 50, 1, 1)
+			mychild.revive(full_heal = 1, admin_revive = 1)
+			if(boosted == 1)
```
if(boosted)

IndieanaJones replied:
Done for upcoming fix.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+		if(core.preserved == 1)
+			visible_message("<span class='boldwarning'>As [user] drops the core into the crevice, the red light intensifies for a brief moment, then returns to normal.</span>")
+			boosted = 1
+			qdel(core)
```
return TRUE here.

IndieanaJones replied:
Resolved in upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+		var/obj/item/organ/regenerative_core/core = W
+		if(core.preserved == 1)
+			visible_message("<span class='boldwarning'>As [user] drops the core into the crevice, the red light intensifies for a brief moment, then returns to normal.</span>")
+			boosted = 1
```
boosted = TRUE

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+/obj/structure/elite_crevice/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/organ/regenerative_core) && activity == 0 && boosted == 0)
+		var/obj/item/organ/regenerative_core/core = W
+		if(core.preserved == 1)
```
if(core.preserved)

IndieanaJones replied:
Done for upcoming fix.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+				newwall = new /obj/effect/temp_visual/elite_crevice_wall(t, src)
+				newwall.activator = src.activator
+				newwall.ourelite = src.mychild
+		sleep(100)
```
timers

IndieanaJones replied:
Done for upcoming fix.

PKPenguin321 replied:
You should know that you don't have to comment on everything like this. When you change the file in a commit, git should automatically mark addressed comments as resolved (or you can do it yourself manually). Commenting can spam notifications

IndieanaJones replied:
Haha, thanks for the heads up.

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,305 @@
+			if(loc)
+				activator.forceMove(loc)
+				visible_message("<span class='boldwarning'>[activator] suddenly reappears above the crevice!</span>")
+				playsound(loc,'sound/effects/phasein.ogg', 200, 0, 50, 1, 1)
```
Inb4 actionninja complains about booleans in playsound.

IndieanaJones replied:
Sure.

ATH1909 commented:
Hell yeah, this sounds super cool. I can't wait to perform a sacrifice ritual to the volca- er, the crevice as an ash walker to gain the favor/protection of a sapient(,) tough lavaland mob/nest mascot (this isn't a criticism of that possibility; I quite like the flavor of it).

> Are not a part of the mining faction, thus being attacked by other fauna

If I might ask, why'd you choose to make this design decision? Why not let them be attacked by other fauna (or maybe give these mini-bosses a toggle-able option that'd let them give and remove the "mining" faction to/from themselves pretty much at-will (with them defaulting to starting with the "mining" faction))?

necromanceranne commented:
Sounds like some Path of Exile shit. I'm super into this.

ARPG loot when?

ATH1909 commented:
How does this interact with a group of people trying to kill this mob together? Can you get it to aggro onto someone who's outside of the barrier, then just snipe it with attacks that can go through walls/barriers?

IndieanaJones commented:
> 
> 
> How does this interact with a group of people trying to kill this mob together? Can you get it to aggro onto someone who's outside of the barrier, then just snipe it with attacks that can go through walls/barriers?

Due to how this is designed, literally *anything* but the two mobs can go through the barrier, including projectiles.  If dippy egg one who isn't the activator stands on the outside of the wall and shoots in, for example, the elite can return in kind by having its projectiles also go through the wall.

ATH1909 commented:
Maybe

> > How does this interact with a group of people trying to kill this mob together? Can you get it to aggro onto someone who's outside of the barrier, then just snipe it with attacks that can go through walls/barriers?
> 
> Due to how this is designed, literally _anything_ but the two mobs can go through the barrier, including projectiles. If dippy egg one who isn't the activator stands on the outside of the wall and shoots in, for example, the elite can return in kind by having its projectiles also go through the wall.

Ah, okay.

Have you considered giving sapient versions of the miniboss a 10-15 second or so grace period of invincibility so that they don't get ganked while learning their controls/reading their flavor text?

IndieanaJones committed:
Cleaned up Code with help from reviewers

IndieanaJones commented:
> 
> 
> Maybe
> 
> > > How does this interact with a group of people trying to kill this mob together? Can you get it to aggro onto someone who's outside of the barrier, then just snipe it with attacks that can go through walls/barriers?
> > 
> > 
> > Due to how this is designed, literally _anything_ but the two mobs can go through the barrier, including projectiles. If dippy egg one who isn't the activator stands on the outside of the wall and shoots in, for example, the elite can return in kind by having its projectiles also go through the wall.
> 
> Ah, okay.
> 
> Have you considered giving sapient versions of the miniboss a 10-15 second or so grace period of invincibility so that they don't get ganked while learning their controls/reading their flavor text?

The spawning currently gives an extra 10 seconds of wait time so the ghost can read the message sent to them explaining the controls and what to do.  I could add invincibility time, though I held against it at the start due to possible abuse.

IndieanaJones commented:
> Sounds like some Path of Exile shit. I'm super into this.
> 
> ARPG loot when?

If this is deemed "okay" to go forward, I would be interested in adding different loot outside of the current pool from the necropolis, perhaps some based on the elites themselves.

IndieanaJones commented:
Every addtimer I just added is very bad, fixing now.

CRITAWAKETS commented:
Honestly this reminds me of the pillar men from Jojo for some reason.

IndieanaJones commented:
So what I'm seeing is that you can't use addtimer like sleep at all, and it requires everything be made into different procs?  God, that seems awful, but alright.

OverlyGeneric commented:
The elite goliath seems kind of bland flavor-wise. 

The fungus attack is unique but overall it doesn't sound very different from the ancient goliath.

Maybe have the sprite of it be infected with fungus from the depths of the Crevice to give more leeway in designing attacks and **lore** reason to why it's being attacked by fellow lavalanders.

Apart from that, I'll throw out some ideas for the elite watcher and elite goldgrub.

**Bloodshot Watcher**

This watcher has been corrupted by the same influence that made the blood-drunk miner.

The combat abilities would involve the watcher using bleed-inducing slashing attacks, gushing blood out like a projectile with properties of capsaicin, warping to a random tile of bloodied ash, and bleeding at low health.

I would imagine this fight to rely on speed and quick reflexes rather than a long dragged out fight.

**Greedle**

A fully-grown goldgrub corrupted by the ruins of Greed and is constantly seeking out more valuables, at any cost, despite no longer needing to consume it.

The combat abilities would involve the beetle disarming its target of their weapons, blinding them with pure greed, the ability the switch forms based on the flip of a coin with a high-damage but slow rhinoceros beetle form and a fast-moving but weak tiger beetle.

I would believe this fight would involve seizing opportunities from both sides and taking advantage of anything in order to win. 

The beetle would run around in their "tails" form disarming and stunning the activator to keep trying their luck to get the "heads" form.

Please remember it's just a suggestion and you don't have to follow it. Thank you for reading this comment.

IndieanaJones commented:
> 
> 
> The elite goliath seems kind of bland flavor-wise.
> 
> The fungus attack is unique but overall it doesn't sound very different from the ancient goliath.
> 
> Maybe have the sprite of it be infected with fungus from the depths of the Crevice to give more leeway in designing attacks and **lore** reason to why it's being attacked by fellow lavalanders.
> 
> Apart from that, I'll throw out some ideas for the elite watcher and elite goldgrub.
> 
> **Bloodshot Watcher**
> 
> This watcher has been corrupted by the same influence that made the blood-drunk miner.
> 
> The combat abilities would involve the watcher using bleed-inducing slashing attacks, gushing blood out like a projectile with properties of capsaicin, warping to a random tile of bloodied ash, and bleeding at low health.
> 
> I would imagine this fight to rely on speed and quick reflexes rather than a long dragged out fight.
> 
> **Greedle**
> 
> A fully-grown goldgrub corrupted by the ruins of Greed and is constantly seeking out more valuables, at any cost, despite no longer needing to consume it.
> 
> The combat abilities would involve the beetle disarming its target of their weapons, blinding them with pure greed, the ability the switch forms based on the flip of a coin with a high-damage but slow rhinoceros beetle form and a fast-moving but weak tiger beetle.
> 
> I would believe this fight would involve seizing opportunities from both sides and taking advantage of anything in order to win.
> 
> The beetle would run around in their "tails" form disarming and stunning the activator to keep trying their luck to get the "heads" form.
> 
> Please remember it's just a suggestion and you don't have to follow it. Thank you for reading this comment.
This stuff seems pretty cool.  The elite goliath in terms of design is a basic bitch, I'll admit, but this mostly stems from my ability to sprite.  However, I do like these ideas, with some tweaking I think they'd fit very well (I have a design plan for the unfinished elites).  I'll see how it turns out.

IndieanaJones committed:
Getting Jiggy with timers

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -104,86 +108,83 @@ While using this makes the system rely on OnFire, it still gives options for tim
 				newwall.activator = src.activator
 				newwall.ourelite = src.mychild
-		sleep(100)
+		addtimer(100)
```
This is not how addtimer works.
```
addtimer(CALLBACK(src, .proc/arena_trap), 10 SECONDS)
addtimer(CALLBACK(src, .proc/border_check), 10 SECONDS)
```

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite_goliath.dm:
```diff
@@ -1,26 +1,8 @@
+  * # Elite Goliath
+  *
+  * An elite mob which boils down to a simple test of keeping a constant eye on the foe and your surrondings.
+  */
```
Why did you remove all the text?
There's nothing wrong with more documentation.

IndieanaJones replied:
How would I go about adding the extra lines?  Would I just add more asterisks?

nemvar commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,311 @@
+				
+obj/structure/elite_crevice/proc/spawn_elite(var/mob/dead/observer/elitemind)
+	var/selectedspawn = pick(potentialspawns)
+	mychild = new selectedspawn(loc, src)
```
Try to avoid sending extra arguments in new(). You don't use the second one, so just remove it.

Mey-Ha-Zah commented:
I have a question. Will these new mini-bosses have unique trophies for Kinetic Crusher?

necromanceranne commented:
That seems like something that'd beed to get added later to would need to be atomized anyway because that then affects crusher balance.

Mey-Ha-Zah commented:
Depending on the effects. Yes Crusher is powerful, but most of its powerful components are locked behind terrible spawn rates (Magmawing/Icewing), or rebalanced boss fights (Ash Drake/Bubblegum [These guys rofl stomped me when I tried to crusher fight them]). So maybe if the boosts aren't too great, or are more thematic in theme (Changing parts of the Crusher's sprite for instance), they can be implemented without affecting balance too much.

IndieanaJones committed:
Update lavaland_surface_elite_crevice.dmm with MapMerger

optimumtact commented:
how does this address lavaland being wildly scaled compared to the station

IndieanaJones commented:
Err...  It doesn't?  The antag freeze was enacted in order for new content to be added, and this is what I'd deem as new content.  While I understand balance is of important concern, nothing here is truly more powerful than what could be already achieved on Lavaland, and a lot of people seem excited about the prospect of what stuff you could do with this. I don't see why it's an issue in this case.

Tlaltecuhtli commented:
no pictures?

IndieanaJones commented:
> 
> 
> no pictures?

Picture will be coming soon after a few tweaks, as a spriter has reached out to me in interest on helping with this PR.  This should work for now, though:
![crevice](https://user-images.githubusercontent.com/47086570/65300024-f9177f00-db3f-11e9-92bf-9ec701043d00.png)

IndieanaJones committed:
Update actions_elites.dmi for Pandora

IndieanaJones committed:
Add pandora.dmi

IndieanaJones committed:
Remove Old Icons File

IndieanaJones committed:
Update elite.dm, elite_goliath.dm.  Add pandora.dm

IndieanaJones committed:
Load Pandora.dm

IndieanaJones commented:
You opened this box, and you better be ready for what comes out.
![pandora](https://user-images.githubusercontent.com/47086570/65365646-2d934580-dbe9-11e9-8e6e-722bc75ad690.png)

Timonkeyn commented:
The crevice looks like a recoloured single tile lava sprite

IndieanaJones commented:
> 
> The crevice looks like a recoloured single tile lava sprite

That's because it's a recolored single tile chasm sprite, which is also just a recolored lava sprite.  I can have the spriter take a look at getting something nicer, but I make no guarantees.

IndieanaJones commented:
Let me tell you, some of these elites will send shivers down your spine.
![legion](https://user-images.githubusercontent.com/47086570/65380727-45350180-dcaf-11e9-99f7-27e6a4f4ef68.png)
Legion is now added to the PR and information on how he works is listed above.
Picture on station:
![legionstation](https://user-images.githubusercontent.com/47086570/65380730-5b42c200-dcaf-11e9-9ffc-f5b59c768207.png)

IndieanaJones committed:
Update actions_elites.dmi

IndieanaJones committed:
Add legionnaire.dmi

IndieanaJones committed:
Update Old Files, add legionnaire.dmi

IndieanaJones committed:
Load legionnaire.dm

IndieanaJones committed:
Update All Files, Add herald.dm

IndieanaJones committed:
Update actions_elites.dmi

IndieanaJones committed:
Update legionnaire.dmi, Add legionnaire_bonfire.dmi and herald.dmi

IndieanaJones committed:
Load herald.dmi

IndieanaJones commented:
DIE.
![herald](https://user-images.githubusercontent.com/47086570/65552565-ba444900-def2-11e9-9410-bc6c3a609fe4.png)
Information on the herald lavaland elite is now in the PR description.
![herald_station](https://user-images.githubusercontent.com/47086570/65552586-caf4bf00-def2-11e9-9040-ab7374cff81b.png)
Other changes came in the form of a different version of the legionnaire's death sprite, and replacing his bone pile with a legion-themed bonfire.  The bonfire acts the same as the bone pile, but flammable mobs which walk over it will be set on fire.

Abductkaz commented:
First time Legionaaire feedback: 250ms ping edition

One idea might be to make the crevice empowered 50% of the time. More ghost roles are always good, and also make sure its really obvious when its empowered (Red lightning? Purple color? Flavor text, of course.)

The head might do to be 2-shotted by the Pka instead of 1shotted like the regular heads, since regular legion spawn loads of them and legionaaire only has a max of 1

Couldn't really find an immediate use for the bonfire on my first go as skele-man, can see how teleporting can be useful after the fact.

More obvious display of current health is needed, I didn't even notice whatever is currently implemented. Maybe the same thing as blobbynauts have but themed around lavaland/each megafauna?

The smoke seems fine, though I was hampered by a small cooldown between selecting an ability and actually using it which made the skull almost never spew smoke since it dies quick, this goes for all the abilities but I assume its there for balance reasons.

Crevice could be more obvious but thats out of your control as a coder anyway.

dee-dubya commented:
Description says that you shouldn't be able to damage yourself, but the Herald can do so with his mirror. If you hit yourself with any bolt from your own mirror(can happen very easily), you take damage.

OverlyGeneric commented:
hoo boy does that legionnaire look **lanky**

sort of looks good but you can't really do much to improve a skeleton sprite.

IndieanaJones commented:
> Description says that you shouldn't be able to damage yourself, but the Herald can do so with his mirror. If you hit yourself with any bolt from your own mirror(can happen very easily), you take damage.
This is due to a bug in the testmerge where Herald's mirror doesn't copy his faction tags should be revived with the crevice shard.  This bug has already been fixed on the internal version.  While it'll say you were hurt by the death bolt in other circumstances, you won't take any damage.

IndieanaJones committed:
Delete elite_goliath.dm

IndieanaJones committed:
Update elite.dm, pandora.dm, legionnaire.dm, and herald.dm.  Add goliath_broodmother.dm

IndieanaJones committed:
Delete lavaland_surface_elite_crevice.dmm

IndieanaJones committed:
Add lavaland_surface_elite_tumor.dmm

IndieanaJones committed:
Update lavaland.dm

IndieanaJones committed:
Delete elite_lavaland_monsters.dmi

IndieanaJones committed:
Delete herald.dmi

IndieanaJones committed:
Delete legionnaire.dmi

IndieanaJones committed:
Delete legionnaire_bonfire.dmi

IndieanaJones committed:
Delete pandora.dmi

IndieanaJones committed:
Add lavaland_elites.dmi

IndieanaJones committed:
Add tumor.dmi and legionnaire_bonfire.dmi

IndieanaJones committed:
Update actions_elites.dmi

IndieanaJones committed:
Stop loading elite_goliath.dm, start loading broodmother.dm

IndieanaJones committed:
Load goliath_broodmother.dm, not broodmother.dm

IndieanaJones commented:
This was the first update since the Testmerge on Sybil, and the changes made reflect on the feedback received, along with the planned rework of the elite goliath.  Let's go down the list so we get everything, shall we?

PULSATING TUMOR:
This is the new, updated version of the glowing crevice.  It behaves identically to the glowing crevice, and just has different visuals and descriptions.
![new_tumors](https://user-images.githubusercontent.com/47086570/66018797-8ba80e80-e4ae-11e9-854b-c13be472d6b8.png)
There have been some updates to the functionality of the tumors from their predecessors, based on feedback:
- Callback time for summoning a fighter back into the arena is now at most 5 seconds, down from 10.
- Different visual forms based on current state (left in the image is unboosted, right is boosted.  Boosted also produces more light)
- Animated sprite (They pulse!).  This makes them more noticable, along with their new appearance.

NEW ELITE: GOLIATH BROODMOTHER
This new elite comes in to replace the old elite goliath that was in previously.  Unlike the old one, this has a custom sprite, and will hopefully fare much better in combat than the old elite.
![broodmother](https://user-images.githubusercontent.com/47086570/66019015-72ec2880-e4af-11e9-9a04-742a240d2649.png)
Information about the Goliath Broodmother can be found up above where the elite goliath's info used to be.
![station_broodmother](https://user-images.githubusercontent.com/47086570/66019086-c9596700-e4af-11e9-8f1d-7774c8bd78df.png)
Along with the new elite, the other elites have received balance adjustments based on the testmerge:

Pandora:
- Linear shot's cooldown has been cut in half (2 sec -> 1 sec).  This is due to linear shot not having as much utility as Pandora's other moves, so the cooldown decrease helps it prove to be viable in more situations.
Legionnaire:
- Skull health increased from 60 to 80.  While not a huge buff, it'll help prevent weaker attacks from oneshotting the skull.
Herald:
- Mirror HP increased from 20 to 60.  This is due to experienced players just smashing Herald's mirror before he could ever use it.  Now, it's a good bit stronger, so you're more likely to get at least one use out of it.

There were also various bugfixes patched in with this update, such as Herald's attack descriptions being inaccurate and him being able to attack his mirror and vice versa.  The PR is almost ready for launchtime, with the next update bringing some of the final features in before its ready.  These features are mainly custom loot drops and a custom HUD for elites, but other small things may also be added.  Til' next time.

NoxVS commented:
Would you be opposed to making it so lava/chasms within the boundaries are made into normal ground? The elites can cross lava so they can drag stuff over there or just hide in lava and be untouchable if you dont have the range

Mey-Ha-Zah commented:
Well, then you maybe shouldn't have made the one near lava sentient. And if that's the case well. Your SOL, sometimes game says, you die.

IndieanaJones committed:
Update all Elite Files for Loot Drops

IndieanaJones committed:
Add elite_trophies.dmi

IndieanaJones committed:
Add hope sprites to accessories.dmi

IndieanaJones committed:
Update neck.dmi

IndieanaJones committed:
Update elite.dm

IndieanaJones committed:
Update screen_objects.dm

IndieanaJones committed:
Add lavaland_elite.dm

IndieanaJones committed:
Add screen_elite.dmi

IndieanaJones committed:
Load Lavaland Elite HUD

IndieanaJones committed:
Update actions_elites.dmi

IndieanaJones commented:
The final updates before I set the PR to Ready.  Adds unique loot for each lavaland elite, which can only be gotten by fighting an elite from a boosted crevice.  You'll have a 50/50 chance to either get the unique loot for that elite (they're listed above) or the tumor shard.  The elites now also have a basic health hud, really nothing special, but it should alert you when you're hurt.  Also applies the red overlay when below certain health thresholds as well.

Taylorkills2 commented:
Bug report with this, if the elite falls into a bottomless pit, it doesn't count as dead. The red barrier won't fall and the activator will be stuck inside. You can repeatedly here the teleportation sound as if the elite was being teleported back.
Can be fixed if an admin kills and revives you in-game.

IndieanaJones committed:
Change the Death-Detection to be Tumor-Based, not Elite-Based

IndieanaJones commented:
> Bug report with this, if the elite falls into a bottomless pit, it doesn't count as dead. The red barrier won't fall and the activator will be stuck inside. You can repeatedly hear the teleportation sound being played as if the elite was being teleported back.
> Can be fixed if an admin kills and revives you in-game.
Thanks for the bug report, just fixed that issue now.

IndieanaJones committed:
Fixing a few rare runtimes

IndieanaJones committed:
Update herald.dm to comply with the obj/projectile change

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+		if(boosted)
+			var/lootpick = rand(1, 2)
+			if(lootpick == 1 && mychild.loot_drop != null)
+				lootbox.contents += new mychild.loot_drop(lootbox.loc)
```
I'm not sure what you tried to achieve because it's pretty mangled but if you just want item inside the lootbox
`new mychild.loot_drop(lootbox)`
same thing below

AnturK reviewed (changes requested):
Just the common stuff for now, will read over the rest tomorrow

IndieanaJones commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+		return
+
+/obj/structure/elite_tumor/proc/arena_checks()
+	if(src) //Checking to see if we still exist
```
QDELETED check is necessary since without it, if the elite or activator falls into a chasm on lavaland, things will break.  The ones in border_check make sure we don't play the teleportation noise and sound when one of them has been QDELETED by the chasm, and the ones in fighters_check make sure the fight ends if someone falls into a chasm.

IndieanaJones commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+/mob/living/simple_animal/hostile/asteroid/elite/Life()
+	. = ..()
+	if(isturf(loc))
+		for(var/obj/structure/elite_tumor/tumor in loc)
```
Yeah, and with it on the structure, the elite's myparent variable can get thrown out, since it was the only thing it was still using it for.

IndieanaJones committed:
Ups the cost of the tumor ruin from 0 to 5

IndieanaJones committed:
Implement fixes suggested by AnturK

IndieanaJones committed:
Compliance to the new curazzy mob verb system

IndieanaJones committed:
Recompliance the whole projectile change thing

IndieanaJones commented:
Thanks for the notes so far, everything you brought up should be addressed.

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/goliath_broodmother.dm:
```diff
@@ -0,0 +1,240 @@
+	visible_message("<span class='warning'>[src] starts picking up speed!</span>")
+	color = rgb(150,0,0)
+	src.set_varspeed(0)
+	src.move_to_delay = 3
```
No need for src

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/goliath_broodmother.dm:
```diff
@@ -0,0 +1,240 @@
+/mob/living/simple_animal/hostile/asteroid/elite/broodmother/proc/reset_rage()
+	color = rgb(255, 255, 255)
+	src.set_varspeed(2)
+	src.move_to_delay = 5
```
Same here

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm:
```diff
@@ -0,0 +1,255 @@
+	
+/mob/living/simple_animal/hostile/asteroid/elite/herald/proc/become_ghost()
+	icon_state = "herald_ghost"
+	return
```
You get a lot of these dangling returns, could clean them up.

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm:
```diff
@@ -0,0 +1,255 @@
+	eyeblur = 0
+	damage_type = BRUTE
+	pass_flags = PASSTABLE
+	var/mob/living/simple_animal/hostile/hostileshooter = null
```
Why this variable exists ? You can just use the firer from what i can see ?

IndieanaJones replied:
Firer is not guaranteed to be a mob, which breaks the faction check further down below.  However, I have redefined it so the variable refers to a mob now instead of specifically a hostile simple mob.

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm:
```diff
@@ -0,0 +1,255 @@
+	desc = "A cloak which protects you from the heresy of the world."
+	icon = 'icons/obj/lavaland/elite_trophies.dmi'
+	icon_state = "herald_cloak"
+	item_state = ""
```
empty item state ?

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/legionnaire.dm:
```diff
@@ -0,0 +1,296 @@
+
+/obj/effect/temp_visual/dragon_swoop/legionnaire/Initialize()
+	. = ..()
+	src.transform *= 0.33
```
No need for src

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/legionnaire.dm:
```diff
@@ -0,0 +1,296 @@
+				return
+			forceMove(T)
+			playsound(src,'sound/effects/bang.ogg', 200, 1)
+			var/list/hit_things
```
You never initialize this list

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/legionnaire.dm:
```diff
@@ -0,0 +1,296 @@
+		icon_living = "legionnaire_headless"
+		icon_aggro = "legionnaire_headless"
+		visible_message("<span class='boldwarning'>[src]'s head flies off!</span>")
+		var/mob/living/simple_animal/hostile/asteroid/elite/legionnairehead/newhead = new /mob/living/simple_animal/hostile/asteroid/elite/legionnairehead(loc)
```
All these 'spawn child mob and give it our current target' stuff could be extracted into a helper on base type.

IndieanaJones replied:
The only thing about that is there's some case-by-case stuff happening depending on the elite.  For example, the broodmother adds all of her kids to a list, whereas the legionnaire has a reference variable to his head, and his summon mechanic also adjusts how much damage the head does, and herald's mirror doesn't do anything regarding targets.

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/legionnaire.dm:
```diff
@@ -0,0 +1,296 @@
+	
+	
+/obj/structure/legionnaire_bonfire/CanPass(atom/movable/mover, turf/target)
+	if(isliving(mover))
```
Bad, wrong, badwrong. CanPass cannot have any side effects.
Depending on intent you want Crossed or Entered.

AnturK reviewed (changes requested):
More things

IndieanaJones committed:
Fixed Up Elite Files Based on Antur's Suggestions

IndieanaJones committed:
That's why shooter was needed

IndieanaJones commented:
Should have covered everything, then.

AnturK commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm:
```diff
@@ -141,7 +142,7 @@
 	if(isliving(target))
 		var/mob/living/L = target
-		if(firer && firer.faction_check_mob(L))
+		if(shooter && shooter.faction_check_mob(L))
```
What i mean you can just do
```
var/mob/living/simple_animal/hostile/asteroid/elite/E = firer
if(istype(E) && E.faction_check_mob(L))
 blahblah
```

IndieanaJones committed:
Let's get rid of shooter

IndieanaJones commented:
Alright, NOW we should be good.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+		if(faction_check_mob(M))
+			return FALSE
+	. = ..()
+	if(ismineralturf(target))
```
I believe this should be `if(. && ismineralturf(target))`

IndieanaJones replied:
Tested this, turns out it does not work, and this setup is required.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+
+/datum/action/innate/elite_attack/Grant(mob/living/L)
+	if(istype(L, /mob/living/simple_animal/hostile/asteroid/elite))
+		M = L
```
Kinda odd.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	var/boosted = FALSE
+	var/times_won = 0
+	var/mob/living/carbon/human/activator = null
+	var/mob/living/simple_animal/hostile/asteroid/elite/mychild = null
```
You need to clear these references to it GC's properly.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	density = FALSE
+	
+/obj/structure/elite_tumor/attack_hand(mob/user)
+	if(istype(user, /mob/living/carbon/human))
```
use the helper.

Also, I believe `attack_hand` needs to return parent values.

IndieanaJones replied:
Could you explain what you mean here better?

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+						SEND_SOUND(elitemind, sound('sound/effects/magic.ogg'))
+						to_chat(elitemind, "<b>You have been chosen to play as a Lavaland Elite.\nIn a few seconds, you will be summoned on Lavaland as a monster to fight your activator, in a fight to the death.\nYour attacks can be switched using the buttons on the top left of the HUD, and used by clicking on targets or tiles similar to a gun.\nWhile the opponent might have an upper hand with  powerful mining equ ...
+						addtimer(CALLBACK(src, .proc/spawn_elite, elitemind), 100)
+						return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+						visible_message("<span class='boldwarning'>The stirring stops, and nothing emerges.  Perhaps try again later.</span>")
+						activity = TUMOR_INACTIVE
+						activator = null
+						return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+						return
+				else
+					addtimer(CALLBACK(src, .proc/spawn_elite), 30)
+				return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+					SEND_SOUND(mychild, sound('sound/effects/magic.ogg'))
+					to_chat(mychild, "<b>Someone has activated your tumor.  You will be returned to fight shortly, get ready!</b>")
+				addtimer(CALLBACK(src, .proc/return_elite), 30)
+				return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+		mychild.key = elitemind.key
+		mychild.sentience_act()
+	INVOKE_ASYNC(src, .proc/arena_trap)
+	return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	if(boosted)
+		mychild.maxHealth = mychild.maxHealth * 2
+		mychild.health = mychild.maxHealth
+	return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+		mychild.health = mychild.maxHealth
+	return
+		
+/obj/structure/elite_tumor/Initialize()
```
mapload arg

IndieanaJones replied:
Could you explain what you mean here better?

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+				H.color = "#FF0000"
+				mychild.update_health_hud()
+		
+/obj/structure/elite_tumor/attackby(obj/item/W, mob/user, params)
```
better to make a cast var like this end with whatever is last used, in this case `item` -> `I`.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+				newwall.activator = src.activator
+				newwall.ourelite = src.mychild
+		addtimer(CALLBACK(src, .proc/arena_checks), 50)
+		return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	INVOKE_ASYNC(src, .proc/arena_trap)  //Gets another arena trap queued up for when this one runs out.
+	INVOKE_ASYNC(src, .proc/border_check)  //Checks to see if our fighters got out of the arena somehow.
+	INVOKE_ASYNC(src, .proc/fighters_check)  //Checks to see if our fighters died.
+	return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	smooth = SMOOTH_TRUE
+	layer = BELOW_MOB_LAYER
+	var/mob/living/carbon/human/activator = null
+	var/mob/living/simple_animal/hostile/asteroid/elite/ourelite = null
```
clear out refs so it GCs.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	else
+		return TRUE
+		
+/obj/structure/elite_tumor/proc/border_check()
```
why do you jump from having code for the tumor back to the effect, then back to the tumor. Code relevant to the object in question should be grouped together.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+	. = ..()
+	if(istype(target, /mob/living/simple_animal/hostile/asteroid/elite) && proximity_flag)
+		var/mob/living/simple_animal/hostile/asteroid/elite/E = target
+		if(E.stat == DEAD && E.sentience_type == SENTIENCE_BOSS)
```
is there every a time where elite's won't have `SENTIENCE_BOSS`? IF not, then this latter check is kinda redundant since you're only allowing this to work on elites anyway.

IndieanaJones replied:
Once they are revived once with a shard, their sentience type switches so they can get laz'd properly.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+				E.desc = "[E.desc]  However, this one appears appears less wild in nature, and calmer around people."
+				E.sentience_type = SENTIENCE_ORGANIC
+				qdel(src)
+				return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+				return
+			else
+				user.visible_message("<span class='notice'>It appears [E] is unable to be revived right now.  Perhaps try again later.</span>")
+				return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+				return
+		else
+			to_chat(user, "<span class='info'>[src] only works on a dead, sentient lavaland elite.  If this one's been revived before, perhaps try a lazarus injection.</span>")
+			return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,354 @@
+			return
+	else
+		to_chat(user, "<span class='info'>[src] only works on a sentient lavaland elite.</span>")
+		return
```
Trailing return

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/goliath_broodmother.dm:
```diff
@@ -0,0 +1,244 @@
+	addtimer(CALLBACK(src, .proc/reset_rage), 65)
+	
+/mob/living/simple_animal/hostile/asteroid/elite/broodmother/proc/reset_rage()
+	color = rgb(255, 255, 255)
```
use the atom color priority system for most color changes.

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/herald.dm:
```diff
@@ -0,0 +1,252 @@
+	
+/mob/living/simple_animal/hostile/asteroid/elite/herald/mirror/death()
+	if(my_master != null)
+		my_master.my_mirror = null
```
put this in `Destroy` instead of `death`

Fox-McCloud commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/legionnaire.dm:
```diff
@@ -0,0 +1,293 @@
+	
+/obj/structure/legionnaire_bonfire/deconstruct(disassembled)
+	if(myowner != null)
+		myowner.mypile = null
```
should be in `Destroy` as well, in case it gets deleted without being deconstructed

Fox-McCloud reviewed (changes requested):
I have some concerns about this, from a design standpoint.

A good portion of the loot is crusher loot, but this is useless for people wielding kinetics. The typical route for spawning crusher loot is currently 100% consistent; if the mob in question took the majority of damage from a crusher, then you have a chance to get the loot (if it's a normal mob), or guaranteed (if it's a boss mob).

There should probably be some other loot that exists here, and the crusher trophies only drop if you' actually...use the crusher.

I realize crusher loot is attractive because it doesn't scale up miners, on station, but...the loot here doesn't have to be strong; it can be gimmicky/unique and that'll be good enough.

I also have concerns over how common *stackable* armor is becoming with lavaland. We're getting dangerously close to being able to become immune to an entire damage type (I suspect it's already possible).

A berserker/inquisitor suit (or the HECK suit) with a security jumpsuit, bracers, codpiece, pandora's hope, and herald cloak is going to *sometimes* make you immune to some melee hits, depending on where you target/happen to hit....and if you happen to have armor from another source, such as the DNA vault or nanites...then this kick into high gear.

I realize some megafauna have armor penetration, but as @optimumtact has pointed out, before, scaling of miners *on station* is a bit ridiculous, and I doubt we want to have miners who are literally melee immune.

IndieanaJones commented:
I'm willing to nerf whatever.  The loot to me isn't actually the main purpose here, it's the fights that you can run to get better at fighting without being gibbed and also having fights that don't play out the same way every time which is important to me.  I personally think that the ash drake armor, for as easy as it is to get, shouldn't have nearly the insane armor resistances it has now (the main feature should being immune to ash storms), but that's just me.

IndieanaJones committed:
Commit changes suggested by Fox-McCloud

IndieanaJones committed:
Forgot a )

IndieanaJones committed:
Remove a H.

IndieanaJones commented:
Alright, that should be everything you suggested.  Now, as for "problematic loot", I think the crusher loot in the boxes isn't too concerning (you can always give it to someone who does use a crusher), though it could be changed in the future.  As for the issue of armor stacking, I'm just going to give them different effects, because I also hate how high miner's armor values can be.  I should have it done in a bit.

IndieanaJones committed:
Update elite.dm, herald.dm, and pandora.dm

IndieanaJones committed:
Add Hope's positive mood effect

IndieanaJones committed:
Consider neck slot item when calculating armor and for block chance

IndieanaJones committed:
Merge branch 'master' into master

IndieanaJones committed:
Fix human_defense.dm

IndieanaJones committed:
Fix Inconsistent Indentation

IndieanaJones committed:
REALLY Fixin' that Indentation

IndieanaJones commented:
Herald and Pandora's loot have both lost their armor-granting abilities and have gained the following:
- Pandora's loot, Hope, now gives a positive moodlet when pinned on your uniform.  This is as strong as the clown's moodlet when he wears his shoes.
- Herald's loot, cloak of the prophet, has a 10% chance to send death bolts in all cardinal and diagonal directions when hit.  These bolts are the same as herald's, doing 15 physical damage with high pierce, and doing 15 item damage when hitting an object.  This effect does have a noise and message in chat before it triggers, giving anyone nearby a second in order to step out of the way.

IndieanaJones commented:
And it errored, but it seems like #47165 will solve that issue.

IndieanaJones committed:
Had to find some obscure thing to fix so Travis would check us out

Mey-Ha-Zah commented:
To quote King Julien: "HOW LONG IS THIS GOING TO TAKE!"

IndieanaJones committed:
While we're still here, fix this off-pixel

IndieanaJones committed:
Background icon fix

IndieanaJones committed:
Merge branch 'master' into master

IndieanaJones committed:
This somehow got left out during the changes

IndieanaJones commented:
I noticed in a final test after all these cleanup changes that the hud wasn't updating properly, so I fixed it.  I also added proper background stuff to the UI buttons.  PR is good for either more review or merge.

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+	faction = list("boss")
+	robust_searching = TRUE
+	ranged_ignores_vision = TRUE
+	ranged = 1
```
This is a book right? Use defines

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+		switch(healthpercent)
+			if(100 to INFINITY)
+				hud_used.healths.icon_state = "elite_health0"
+			if(80 to 100)
```
Only set severity in this switch then below just do icon_state = "elite_health[severity]"

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+				var/mob/dead/observer/elitemind = null
+				visible_message("<span class='boldwarning'>The tumor begins to convulse.  Your instincts tell you to step back.</span>")
+				activator = user
+				if(boosted)
```
This would read much nicer if you did.!boosted then add the timer and return. Then you don't have to nest so deep

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+	. = ..()
+	if(istype(I, /obj/item/organ/regenerative_core) && activity == TUMOR_INACTIVE && !boosted)
+		var/obj/item/organ/regenerative_core/core = I
+		if(core.preserved)
```
Early return instead

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+			
+/obj/structure/elite_tumor/proc/arena_trap()
+	var/turf/T = get_turf(src)
+	if(T && activity == TUMOR_ACTIVE)
```
Bad nesting

Qustinnus commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/elites/elite.dm:
```diff
@@ -0,0 +1,352 @@
+		
+/obj/structure/elite_tumor/proc/border_check()
+	if(activity == TUMOR_ACTIVE)
+		if(activator != null && !QDELETED(activator) && get_dist(src, activator) >= 12)
```
You get the idea, you nest far too deep do a check on the entire pr yourself you nest deep all over it.

IndieanaJones replied:
OK

IndieanaJones committed:
Update to elite files

IndieanaJones committed:
Merge branch 'master' into master

IndieanaJones commented:
Fixed the nesting and removed some unnecessary checks.  Also organized some of the elements better, such as each individual mob/effect not being in the middle of their elite's code, but at the end instead.  Also optimized the arena checks a bit better, and moved the looping part to the main body instead of in the barrier method.

IndieanaJones committed:
Update the Description of the Herald

IndieanaJones commented:
This was a small change in order to go through the checks again since the icon bot broke for the last version.  However, it does change the description of the herald fight to be more accurate to what happens during the fight, so there's that.
